chain/client/src/chunk_validation.rs

```diff
@@ -49,14 +54,20 @@ impl ChunkValidator {
         network_sender: Sender<PeerManagerMessageRequest>,
         runtime_adapter: Arc<dyn RuntimeAdapter>,
     ) -> Self {
-        Self { my_signer, epoch_manager, network_sender, runtime_adapter }
+        Self {
+            my_signer,
+            epoch_manager,
+            network_sender,
+            runtime_adapter,
+            chunks_in_validation: Default::default(),
+        }
     }
 
     /// Performs the chunk validation logic. When done, it will send the chunk
     /// endorsement message to the block producer. The actual validation logic
     /// happens in a separate thread.
     pub fn start_validating_chunk(
-        &self,
+        &mut self,
         state_witness: ChunkStateWitness,
         chain_store: &ChainStore,
     ) -> Result<(), Error> {
```

Comment:
Another way I can think of waiting for completion of all validation is to pass in an optional atomic counter or optional `chunks_in_validation` like structure to the `start_validating_chunk` instead of having it in `ChunkValidator` struct.

Advantage of this are 
- We would only use `chunks_in_validation` for test scenarios. With the current way it's structured, it can technically indefinitely grow.
- The `wait_for_chunks_in_validation` functionality shifts to the integration test instead of being a part of `ChunkValidator` struct.

The call path to start_validating_chunk is 
`env.propagate_chunk_state_witnesses` -> `client.process_chunk_state_witness` -> `chunk_validator.start_validating_chunk`

Reply:
We can potentially even pass in a callback FnOnce to `start_validating_chunk`, call it at the end of validation, and keep the logic of polling, initializing chunks_in_validation etc. all within the test module.

Reply:
That's interesting idea, but I think a hidden test-only field in `ChunkValidator` is better than adding new argument. The future callers may not know about the field, but in case of argument they will have to think if None is fine to pass. Especially callbacks strongly decrease readability.

Reply:
Hmm, here's another idea - perhaps as part of the ChunkValidator struct we can have a
`  validated: Sender<ChunkHash>`
where Sender is from core/async/messaging.rs

It doesn't do anything other than providing a send function. For production we can initialize it with Sender::noop(). For tests, we need some other mechanism to capture the events. For that we can make a struct that implements CanSend (only thing that's needed is to implement the send function) such as this

```rust
struct ValidatedChunksSender {
  sender: std::sync::mpsc::Sender<ChunkHash>,
}

impl near_async::messaging::CanSend<ChunkHash> for ValidatedChunksSender {
  fn send(&self, x: ChunkHash) { self.sender.send(x).unwrap(); }
}

impl ValidatedChunksSender {
  fn new() -> (near_async::messaging::Sender<ChunkHash>, std::sync::mpsc::Receiver<ChunkHash>) {
    let (sender, receiver) = std::sync::mpsc::channel(100);
    let sender = near_async::messaging::Sender::from_impl(Self { sender});
    (sender, receiver)
  }
}

fn wait_for_chunks_to_be_validated(mut chunks: HashSet<ChunkHash>, mut receiver: std::sync::mpsc::Receiver<ChunkHash>) {
  while !chunks.is_empty() {
    let chunk = receiver.recv().unwrap();
    chunks.remove(&chunk);
  }
}

fn test_blah() {
  ...
  let mut chunks = HashSet::new();
  let (sender, mut receiver) = ValidatedChunksSender::new();
  let mut chunk_validator = ChunkValidator {
    ...
    validated: sender
  };
  for any chunk to validate {
    chunks.insert(chunk);
  }
  wait_for_chunks_to_be_validated(chunks, receiver);
}
```

(Even if we don't use near_async::messaging::Sender, we can still consider using std::sync::mpsc::Sender - the practical difference is it doesn't allow a noop implementation. Still, having the receiver side recv() instead of having an arc of a lock of hashmap of arc of oncecells this is simpler.)